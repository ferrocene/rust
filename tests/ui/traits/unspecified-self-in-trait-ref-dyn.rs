//@ edition: 2018..

pub trait Foo<A=Self> {
    fn foo(&self);
}

pub trait Bar<X=usize, A=Self> {
    fn foo(&self);
}

fn main() {
    let a = <dyn Foo>::lol();
    //~^ ERROR must be explicitly specified
    let b = <dyn Foo<_>>::lol();
    //~^ ERROR no function or associated item named
    let c = <dyn Bar>::lol();
    //~^ ERROR must be explicitly specified
    let d = <dyn Bar<usize, _>>::lol();
    //~^ ERROR no function or associated item named
    let e = <dyn Bar<usize>>::lol();
    //~^ ERROR must be explicitly specified
}
