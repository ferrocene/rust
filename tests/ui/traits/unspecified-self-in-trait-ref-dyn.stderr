error[E0393]: the type parameter `A` must be explicitly specified
  --> $DIR/unspecified-self-in-trait-ref-dyn.rs:12:18
   |
LL | pub trait Foo<A=Self> {
   | --------------------- type parameter `A` must be specified for this
...
LL |     let a = <dyn Foo>::lol();
   |                  ^^^
   |
   = note: because the parameter default references `Self`, the parameter must be specified on the object type
help: set the type parameter to the desired type
   |
LL |     let a = <dyn Foo<A>>::lol();
   |                     +++

error[E0599]: no function or associated item named `lol` found for trait object `dyn Foo<_>` in the current scope
  --> $DIR/unspecified-self-in-trait-ref-dyn.rs:14:27
   |
LL |     let b = <dyn Foo<_>>::lol();
   |                           ^^^ function or associated item not found in `dyn Foo<_>`

error[E0393]: the type parameter `A` must be explicitly specified
  --> $DIR/unspecified-self-in-trait-ref-dyn.rs:16:18
   |
LL | pub trait Bar<X=usize, A=Self> {
   | ------------------------------ type parameter `A` must be specified for this
...
LL |     let c = <dyn Bar>::lol();
   |                  ^^^
   |
   = note: because the parameter default references `Self`, the parameter must be specified on the object type
help: set the type parameter to the desired type
   |
LL |     let c = <dyn Bar<A>>::lol();
   |                     +++

error[E0599]: no function or associated item named `lol` found for trait object `dyn Bar<usize, _>` in the current scope
  --> $DIR/unspecified-self-in-trait-ref-dyn.rs:18:34
   |
LL |     let d = <dyn Bar<usize, _>>::lol();
   |                                  ^^^ function or associated item not found in `dyn Bar<usize, _>`

error[E0393]: the type parameter `A` must be explicitly specified
  --> $DIR/unspecified-self-in-trait-ref-dyn.rs:20:18
   |
LL | pub trait Bar<X=usize, A=Self> {
   | ------------------------------ type parameter `A` must be specified for this
...
LL |     let e = <dyn Bar<usize>>::lol();
   |                  ^^^^^^^^^^ missing reference to `A`
   |
   = note: because the parameter default references `Self`, the parameter must be specified on the object type

error: aborting due to 5 previous errors

Some errors have detailed explanations: E0393, E0599.
For more information about an error, try `rustc --explain E0393`.
